### PR TITLE
Remove support for dynamic plugin imports

### PR DESCRIPTION
**Pull request by Viicos**
This seems to be a blocker for Manim to work under 3.12.

This PR only focuses on fixing the issue with `pkg_resources`, but is this feature really wanted? It seems to achieve the following: https://github.com/ManimCommunity/manim/pull/967#issuecomment-769026805

Honestly, I don't see how this is beneficial: IDEs/linters are not aware of the import (see screenshot). I'm in favor of removing this dynamic import (plus having these `exec` calls is scary, seems like you can have arbitrary code execution by crafting a specific module name).

### CODE REVIEW

**Comment by jsonvillanueva** [comment]
> Honestly, I don't see how this is beneficial: IDEs/linters are not aware of the import (see screenshot). I'm in favor of removing this dynamic import (plus having these exec calls is scary, seems like you can have arbitrary code execution by crafting a specific module name).

On the one hand, I agree.. the exec is a bit scary. On the other, I *could see* a world where the dynamic import becomes more useful (e.g. UI editor). I'm indifferent to its removal, but perhaps @naveen521kk could chime in.

**Comment by Viicos** [comment]
Hi @jsonvillanueva, I see you are the one who posted the screenshot from the comment I've linked. When writing code in an editor that does not provide LSP features (such as Vanilla vim, or VSCode without Pylance), doing `from manim import *` and using the provided plugin objects (classes, functions, whatever) might save you an import. But for anyone else, the IDE will go crazy because you are using an undefined object (so no autocompletion, goto, etc).

Having this dynamic import is highly unusual and I don't see any benefit of saving one import over having no info about the imported objects whatsoever.

**Comment by MrDiver** [comment]
Can we just import it behind if TYPE_CHECKING?

**Comment by Viicos** [comment]
> Can we just import it behind if TYPE_CHECKING?

By _it_ do you mean the plugin object used? If so, there's no point in having it dynamically imported by manim. Importing the object yourself would actually save you a `if TYPE_CHECKING` statement

**Comment by behackl** [comment]
I think it might be worth it to remove this particular functionality. From most plugins that I've seen, none really made use of this -- and even if they did, it probably is an antipattern that we shouldn't really reinforce. Linters and IDEs in general are unhappy with automated imports and namespace assignments.

**Comment by jsonvillanueva** [comment]
> When writing code in an editor that does not provide LSP features (such as Vanilla vim, or VSCode without Pylance), doing `from manim import *` and using the provided plugin objects (classes, functions, whatever) might save you an import. But for anyone else, the IDE will go crazy because you are using an undefined object (so no autocompletion, goto, etc).

In my opinion, this dynamic plugin loading feature is only useful in a non-coding, user facing context. This context guarantees the user is NOT writing code. This is a bit odd, given that Manim is generally meant for people who read and write python code, but I'm specifically imagining use-cases with Dear ImGUI, or other non-code editors like Manim Studio.

The BIG caveat here is that we're far from a finished OpenGL rewrite (and thus Dear ImGUI support) and there isn't a stable full fledged non-code editor (like Blender) to take advantage of this -- so again, I'm indifferent to its removal.

**Comment by Viicos** [comment]
> but I'm specifically imagining use-cases with Dear ImGUI, or other non-code editors like Manim Studio.

I don't know really much about these tools, but now that I hear about it, it makes a little bit more sense. However, I think it would be better if the dynamic imports would be implemented in these tools instead, in some way (if not possible in those tools -- as I said, don't know really anything about them -- we could for example gate this feature behind an `if` check with some env variable or similar :))

**Comment by naveen521kk** [comment]
> I'm indifferent to its removal, but perhaps @naveen521kk could chime in

It's been really long since I've worked on this, though if I remember correctly one of the main reason this exists is for plugins which only monkey patches manim and doesn't expose any new objects. If I remember correctly, one of my main use-case back then was some customisation related to default fonts/colors which are monkey patched by a custom plugin which I have written. The plugin name would be in my user-level config file, so every time I don't need to import my plugin. It's a really niche use-case though so I'm not really opposed to removing them.

**Comment by Viicos** [comment]
> If I remember correctly, one of my main use-case back then was some customisation related to default fonts/colors which are monkey patched.

Ah in that case it makes sense as well. But I think in that case we would need a proper plugin structure: instead of blindly importing everything, we should enforce a specific function/class/whatever to be present, so that we can still keep the plugin "autoload feature", without having to do shady stuff with `eval`

File: manim/plugins/__init__.py [hunk 1]
```diff
@@ -1,3 +1,16 @@
 from __future__ import annotations
 
-from .import_plugins import *
+from .. import config, logger
```
Comment (JasonGrace2282) [inline manim/plugins/__init__.py:3]:
```suggestion
from manim._config import config, logger
```
I believe MrDiver and Chopan were trying to move away from these kinds of imports.

File: manim/plugins/__init__.py [hunk 2]
```diff
@@ -1,3 +1,16 @@
...
+from .plugins_flags import get_plugins, list_plugins
```
Comment (JasonGrace2282) [inline manim/plugins/__init__.py:4]: Same as above

**Review by JasonGrace2282 — CHANGES_REQUESTED** [review]
The changes look good to me, just one minor thing about the imports ;)

File: manim/plugins/__init__.py (hunk 1 above)
Comment (Viicos) [reply manim/plugins/__init__.py:3]: True, going with relative imports is a good practice if it's at the same level.

File: manim/plugins/__init__.py (hunk 2 above)
Comment (Viicos) [reply manim/plugins/__init__.py:4]: As said in my previous comment, relative imports are beneficial to some extent (i.e. if not going up too much). It is easier to refactor (if you rename a parent module, you'd have to rename your imports as well if you had used absolute imports).

**Review by naveen521kk — DISMISSED** [review]
Can you update the docs? https://docs.manim.community/en/stable/plugins.html#creating-plugins

**Review by JasonGrace2282 — APPROVED** [review]
LGTM!

File: docs/source/plugins.rst [hunk 3]
```diff
@@ -99,40 +99,18 @@ directory structure, build system, and naming are completely up to your
 discretion as an author. The aforementioned template plugin is only a model
 using Poetry since this is the build system Manim uses. The plugin's `entry
 point <https://packaging.python.org/specifications/entry-points/>`_ can be
-specified in poetry as:
+specified in Poetry as:
```
Comment (Viicos) [inline docs/source/plugins.rst:102]: Self-reminder: a generic example should be given, not necessarily Poetry

**Comment by Viicos** [comment]
@naveen521kk docs updated

**Pull request merged**